### ISSUE TITLE
Use for Symbol-based soft-private

### ISSUE THREAD

**Issue by bakkot** [littledan/proposal-class-static-block#1]
I quite like the idea of

```js
class Foo {
  static { const x = Symbol('x'); }
  [x] = kinda_secret;
  peekInside() {
    return this[x];
  }
  static peekInside(obj) {
    return obj[x];
  }
}
```

Thoughts?

---

As I've written it the `const` declaration would be visible in the whole body of the class, not just in the static block, which might be surprising / inconsistent with the normal visibility rules, but I think it would be ok.

---

This wouldn't work with the proposal as written because the LHS in `[x] = ` would be evaluated before the static block, although this would:

```js
let x;
class Foo {
  static { x = Symbol('x'); }
  constructor() {
    this[x] = kinda_secret;
  }
  peekInside() {
    return this[x];
  }
  static peekInside(obj) {
    return obj[x];
  }
}
```

I'd argue that this inconsistency suggests the static block's evaluation should be interleaved with the evaluation of LHSes of computed property names, which is also the execution order implied by the source order; perhaps there could be a special case for static blocks which were the last (non-`;`) element in a class body which would allow such blocks to access the class binding.

**Comment by zenparsing** [comment]
Interesting. I think I would find it surprising for variable declarations to have visibility outside of the block braces.

Also, the class binding isn't initialized at the time that computed properties are evaluated, but I guess I assumed that the binding would be initialized when the static block runs.

What do you think?

**Comment by zenparsing** [comment]
Oh, I just read your last sentence. I think it would be strange for the position of the static block to make a difference with respect to the class binding.

**Comment by bakkot** [comment]
I agree with all of that. On the other hand, I would also find it surprising if it was executed out of order with computed property names, since those are also code which is evaluated in the class body. That could be resolved by only allowing a static block as the last element in a class, but it does seem useful to allow it to appear and execute earlier, even though in those positions it couldn't access the class binding.

As for its position affecting the availability of the class binding, it would be one more thing to keep track of, but I think "the class binding is not available until the class's structure is complete" makes at least some amount of sense.

Visibility of declarations in the static block is probably another topic entirely. Making them visible in the whole class does have a mental model which makes sense (i.e. code in a static block is executed in the class body, rather than in a block in the class body), but it is different from normal visibility rules, at least with the current syntax.

I guess what I'm actually looking for is static private fields which are available to computed property names. Unfortunately I doubt that's going to happen, but the above would accomplish the same thing. In fact, static private fields would be strictly less useful than having static blocks which could run before computed property names were evaluated and whose declarations were visible throughout the class. Here's a static private "field":

```js
class C {
  static { let instances = 0; }
  static get instances() {
    return instances;
  }
  sequenceNumber = instances++;
}

(new C).sequenceNumber; // 0
(new C).sequenceNumber; // 1
C.instances; // 2
```

If we don't allow this, I think some people are going to be surprised that it doesn't work. But perhaps we shouldn't allow both this and private static fields.
